lightning/src/events/mod.rs

```diff
@@ -2218,15 +2268,38 @@ impl MaybeReadable for Event {
 			25u8 => {
 				let mut f = || {
 					let mut prev_channel_id = ChannelId::new_zero();
-					let mut failed_next_destination_opt = UpgradableRequired(None);
+					let mut failure_reason = None;
+					let mut failure_type_opt = UpgradableRequired(None);
 					read_tlv_fields!(reader, {
 						(0, prev_channel_id, required),
-						(2, failed_next_destination_opt, upgradable_required),
+						(1, failure_reason, option),
+						(2, failure_type_opt, upgradable_required),
 					});
-					Ok(Some(Event::HTLCHandlingFailed {
+
+					let mut event = Event::HTLCHandlingFailed {
 						prev_channel_id,
-						failed_next_destination: _init_tlv_based_struct_field!(failed_next_destination_opt, upgradable_required),
-					}))
+						failure_type: _init_tlv_based_struct_field!(failure_type_opt, upgradable_required),
+						failure_reason,
+					};
+
+					// The [`HTLCHandlingFailureType::UnknownNextPeer`] variant is deprecated, but
+					// we continue writing it to allow downgrading. If it was written, upgrade
+					// it to its new representation of [`HTLCHandlingFailureType::InvalidForward`]
+					// and [`LocalHTLCFailureReason::UnknownNextHop`]. This will cover both the case
+					// where we have a legacy event and new events that are written with the legacy
+					// type be downgradable.
+					match event {
+						Event::HTLCHandlingFailed { failure_type: HTLCHandlingFailureType::UnknownNextHop { requested_forward_scid }, .. } => {
+								event = Event::HTLCHandlingFailed {
+									prev_channel_id,
+									failure_type: HTLCHandlingFailureType::InvalidForward { requested_forward_scid },
+									failure_reason: Some(LocalHTLCFailureReason::UnknownNextPeer.into()),
+								}
+						}
+						_ => panic!("HTLCHandlingFailed wrong type")
```

Comment:
Wait, is this panic reachable? Consider using `if let` instead of `match` to avoid it

Reply:
Yikes yeah, reverted to just match on `HTLCHandlingType` (was like this before [change suggestion](https://github.com/lightningdevkit/rust-lightning/pull/3700#discussion_r2050551890)) - think the panic should be ok if it's for a variant we just declared right above?

Reply:
Yeah  it's fine, although generally find it more readable to just debug_assert(false) and return `Ok(None)` so there's no need to double check the prod panic